# **Auditoría De Calidad De Datos**

Integrantes: Juan Diego Serpa Ruiz, Juan Esteban Camacho Castro, Santiago Rueda Plata, Juan Pablo González Ricaurte

## **1. Introducción**

### **1.1. Objetivo**

Evaluar la calidad de la información en la base de datos de NovaMarket identificando anomalías, incoherencias y demás aspectos que generen una alteración en los datos. El propósito es establecer un plan de saneamiento antes de antes de usarlos en los reportes y modelos de análisis del negocio.




### **1.2. Carga de librerías, dataset y descripción del Dataset**

In [1]:
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns

df = pd.read_csv("NovaMarket_datos_crudos.csv")
df.head()

,id_pedido,id_cliente,fecha_pedido,canal_compra,metodo_pago,ciudad_tienda,codigo_postal,categoria_producto,producto,precio_unitario,unidades_vendidas,monto_compra,edad_cliente,correo_cliente,nivel_satisfaccion,nivel_lealtad,comentario_cliente,peso_pedido_kg,fecha_actualizacion_stock,fuga_cliente
0,P505195,C45650,15/11/2025,Tienda,Efectivo,bogota,1100103,Moda,Chaqueta impermeable,28700.0,1,28700.0,32,cliente405195@novamarket.com.co,Bajo,Platino,El domiciliario fue muy amable,14.9,2026-04-17,0
1,P677836,C84260,2026-03-05,Web,Efectivo,MZL,1700101,Electrónica,Parlante bluetooth,195700.0,2,391400.0,73,cliente577836@yahoo.es,Medio,Bronce,El producto llegó dañado,12.9,2026-06-10,0
2,P387869,C164166,2025-03-13,Tienda,PayPal,Cúcuta,5400101,Juguetería,Rompecabezas 1000 piezas,597600.0,-3,-1792800.0,32,cliente287869@novamarket.com.co,Bajo,Oro,El domiciliario fue muy amable,14.3,2026-02-26,0
3,P481069,C98406,27/06/2025,App,transferencia,Cucuta,5400103,deportes,Colchoneta de yoga,322900.0,1,322900.0,45,cliente381069@gmail.com,Medio,Platino,Rápido y sin problemas,9.5,2026-05-14,0
4,P247920,C163018,01/05/2025,Web,Efectivo,Bucaramanga,6800112,Hogar,Juego de sábanas,222400.0,1,222400.0,60,cliente147920@novamarket.com.co,Alto,Plata,NaN,12.3,2026-07-13,0


<div>

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>id_pedido</th>
      <th>id_cliente</th>
      <th>fecha_compra</th>
      <th>canal</th>
      <th>ciudad</th>
      <th>codigo_postal</th>
      <th>categoria_producto</th>
      <th>producto</th>
      <th>precio</th>
      <th>unidades</th>
      <th>edad_cliente</th>
      <th>correo</th>
      <th>nivel_satisfaccion</th>
      <th>fecha_actualizacion_stock</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>P00383</td>
      <td>C0276</td>
      <td>2026-01-03</td>
      <td>App</td>
      <td>Bucaramanga</td>
      <td>68001</td>
      <td>Belleza</td>
      <td>Camiseta</td>
      <td>-1742611.0</td>
      <td>4</td>
      <td>-19</td>
      <td>c0276gmail.com</td>
      <td>alto</td>
      <td>2026-07-01</td>
    </tr>
    <tr>
      <th>1</th>
      <td>P00403</td>
      <td>C0167</td>
      <td>6/12/2026</td>
      <td>Web</td>
      <td>Cali</td>
      <td>76001</td>
      <td>Deportes</td>
      <td>Cafetera</td>
      <td>1828583.0</td>
      <td>2</td>
      <td>23</td>
      <td>c0167@hotmail.com</td>
      <td>MEDIO</td>
      <td>2026-07-01</td>
    </tr>
    <tr>
      <th>2</th>
      <td>P00403</td>
      <td>C0167</td>
      <td>6/12/2026</td>
      <td>Web</td>
      <td>Cali</td>
      <td>76001</td>
      <td>Deportes</td>
      <td>Cafetera</td>
      <td>1828583.0</td>
      <td>2</td>
      <td>23</td>
      <td>c0167@hotmail.com</td>
      <td>MEDIO</td>
      <td>2026-07-01</td>
    </tr>
    <tr>
      <th>3</th>
      <td>P00450</td>
      <td>C0128</td>
      <td>07/09/2026</td>
      <td>App</td>
      <td>bucaramanga</td>
      <td>68001</td>
      <td>electronica</td>
      <td>Set bloques</td>
      <td>787178.0</td>
      <td>6</td>
      <td>69</td>
      <td>c0128@hotmail.com</td>
      <td>MEDIO</td>
      <td>2026-07-15</td>
    </tr>
    <tr>
      <th>4</th>
      <td>P00449</td>
      <td>C0304</td>
      <td>2025-07-19</td>
      <td>Tienda</td>
      <td>Bogotá</td>
      <td>8001</td>
      <td>Belleza</td>
      <td>Audífonos BT</td>
      <td>357231.0</td>
      <td>5</td>
      <td>67</td>
      <td>c0304@novamarket.co</td>
      <td>Bajo</td>
      <td>2026-05-20</td>
    </tr>
  </tbody>
</table>
</div>




In [2]:
print(df.describe())

print("\n", df.shape)

print("\n", df.columns)

print("\n", df.dtypes)

print("\n", df.info)

       codigo_postal  precio_unitario  unidades_vendidas  monto_compra  \
count   6.400000e+05     6.183600e+05      640000.000000  6.183600e+05   
mean    4.063765e+06     5.498868e+06         219.419042  1.281407e+09   
std     2.667342e+06     7.003612e+07        1330.372612  9.991062e+10   
min     5.001010e+05    -8.881000e+05          -3.000000 -8.580142e+09   
25%     1.100170e+06     1.697000e+05           1.000000  2.226000e+05   
50%     4.700112e+06     3.056000e+05           1.000000  4.565000e+05   
75%     6.600192e+06     4.842000e+05           3.000000  8.568000e+05   
max     7.600195e+06     9.990000e+08        9999.000000  9.989001e+12   

        edad_cliente  peso_pedido_kg   fuga_cliente  
count  640000.000000   640000.000000  640000.000000  
mean       48.818192       68.466064       0.209027  
std        28.740425      250.219075       0.406614  
min        -5.000000        8.000000       0.000000  
25%        31.000000        9.800000       0.000000  
50%      

El dataset tiene 620000 filas y 20 columnas. La mayoría de los campos son de tipo texto, mientras que solo siete columnas son numéricas (codigo_postal, precio_unitario, unidades_vendidas, monto_compra, edad_cliente, peso_pedido_kg, fuga_cliente). A primera vista los datos presentan problemas como: valores nulos, registros duplicados, inconsistencias en el formato de textos (como diferencias entre mayúsculas y minúsculas) y datos anormales como precios, edades y unidades con valores negativos o fuera de rango.



## **2. Diagnóstico por Dimensión**

### **2.1. Completitud**
En esta sección se audita **cuánta información está disponible** en cada columna. Se cubren dos aspectos:

- **Nulos reales (*NaN*)**: celdas explícitamente vacías, que se detectan con `isnull()`.
- **Falsos completos**: celdas que contienen texto como *"N/A"*, *"Sin dato"* o *"?"*, que aparentan estar completas pero significan ausencia de dato. Estas se revisan con `value_counts(dropna=False)` sobre las columnas categóricas más relevantes.

In [27]:
df.isna().sum()

id_pedido                         0
id_cliente                        0
fecha_pedido                      0
canal_compra                      0
metodo_pago                       0
ciudad_tienda                 29928
codigo_postal                     0
categoria_producto                0
producto                          0
precio_unitario               21640
unidades_vendidas                 0
monto_compra                  21640
edad_cliente                      0
correo_cliente               105312
nivel_satisfaccion            46469
nivel_lealtad                     0
comentario_cliente            90806
peso_pedido_kg                    0
fecha_actualizacion_stock         0
fuga_cliente                      0
dtype: int64

In [28]:
(df.isnull().mean() * 100).round(1)

id_pedido                     0.0
id_cliente                    0.0
fecha_pedido                  0.0
canal_compra                  0.0
metodo_pago                   0.0
ciudad_tienda                 4.7
codigo_postal                 0.0
categoria_producto            0.0
producto                      0.0
precio_unitario               3.4
unidades_vendidas             0.0
monto_compra                  3.4
edad_cliente                  0.0
correo_cliente               16.5
nivel_satisfaccion            7.3
nivel_lealtad                 0.0
comentario_cliente           14.2
peso_pedido_kg                0.0
fecha_actualizacion_stock     0.0
fuga_cliente                  0.0
dtype: float64

**Interpretación del recuento de nulos**

Se detectan **6 columnas con valores faltantes**, ordenadas por impacto:

| Columna | Nulos | % del total |
|---|---|---|
| *correo_cliente* | 105.312 | 16.46% |
| *comentario_cliente* | 90.806 | 14.19% |
| *nivel_satisfaccion* | 46.469 | 7.26% |
| *ciudad_tienda* | 29.928 | 4.68% |
| *precio_unitario* | 21.640 | 3.38% |
| *monto_compra* | 21.640 | 3.38% |

Las 14 columnas restantes están completas (0% de nulos). Se observa que *precio_unitario* y *monto_compra* tienen exactamente el mismo número de nulos, lo cual sugiere que faltan en las mismas filas (posiblemente por la relación *monto_compra = precio_unitario × unidades_vendidas*).

A continuación se revisa con `value_counts()` si en las columnas categóricas más importantes existen valores que aparentan estar completos pero corresponden a datos ausentes (falsos completos).

In [37]:
df['nivel_satisfaccion'].value_counts(dropna=False)

nivel_satisfaccion
Alto      166739
Medio     165777
Bajo      165309
NaN        46469
alto       12803
medio      12661
bajo       12585
BAJO        6582
 Alto       6554
 Medio      6444
Medio       6429
 Bajo       6383
MEDIO       6347
Alto        6335
ALTO        6315
Bajo        6268
Name: count, dtype: int64

**Hallazgos – *nivel_satisfaccion***

- Nulos: 46.469 (7.26%)
- Cardinalidad reportada: 15 valores únicos + NaN
- Cardinalidad real: **3 niveles** (Alto, Medio, Bajo)
- Frecuencias agrupadas: Alto ≈ 198.746 (31%) · Medio ≈ 197.658 (31%) · Bajo ≈ 197.127 (31%)
- Inconsistencias detectadas:
  - Mayúsculas/minúsculas: *Alto*/*ALTO*/*alto*, *Medio*/*MEDIO*/*medio*, *Bajo*/*BAJO*/*bajo*.
  - **Espacios al inicio**: aparecen entradas como **Alto**, **Medio**, **Bajo** que Python cuenta como categorías distintas.
- Distribución real bastante balanceada entre los 3 niveles.

In [38]:
df['nivel_lealtad'].value_counts(dropna=False)

nivel_lealtad
Bronce     160420
Oro        160222
Plata      159912
Platino    159446
Name: count, dtype: int64

**Hallazgos – *nivel_lealtad***

- Nulos: 0
- Cardinalidad: 4 valores únicos ✅
- Frecuencias: Bronce = 160.420 · Oro = 160.222 · Plata = 159.912 · Platino = 159.446
- **Columna limpia**: sin nulos, sin inconsistencias, distribución muy balanceada entre los 4 niveles de lealtad.

In [39]:
df['ciudad_tienda'].value_counts(dropna=False)

ciudad_tienda
NaN              29928
Villavicencio    27634
Pereira          27549
Cali             25462
Bucaramanga      25461
                 ...  
 Bogotá           2662
bogota d.c.       2654
BOGOTÁ            2646
BOGOTA DC         2638
BOG               2630
Name: count, Length: 89, dtype: int64

**Hallazgos – *ciudad_tienda***

- Nulos: 29.928 (4.68%)
- Cardinalidad reportada: 89 valores únicos (incluyendo NaN)
- Cardinalidad real: aproximadamente **10-12 ciudades reales**
- Inconsistencias detectadas:
  - Mayúsculas/minúsculas: *Bogotá*/*BOGOTÁ*/*bogota*.
  - Acentos: *Bogota*/*Bogotá*, *Medellin*/*Medellín*.
  - Espacios al inicio: *Bogotá*, *Cali*.
  - Abreviaturas y variantes administrativas: *BOG*, *BOGOTA DC*, *bogota d.c.*
- La cardinalidad de 89 valores está inflada artificialmente por estas variantes de escritura de las mismas ciudades reales.

In [40]:
df['categoria_producto'].value_counts(dropna=False)

categoria_producto
Belleza         56378
Electrónica     56261
Deportes        56019
Hogar           55846
Juguetería      55499
Moda            55300
moda            20354
hogar           20352
deportes        20351
belleza         20273
electronica     10433
JUGUETERÍA      10332
HOGAR           10290
Electrónica     10267
DEPORTES        10264
Deportes        10214
Belleza         10195
Moda            10182
 Juguetería     10170
Hogar           10157
 Deportes       10147
electrónica     10142
 Moda           10136
juguetería      10133
BELLEZA         10114
ELECTRÓNICA     10085
 Belleza        10052
Juguetería      10047
jugueteria      10029
MODA            10025
 Electrónica    10008
 Hogar           9945
Name: count, dtype: int64

**Hallazgos – *categoria_producto***

- Nulos: 0
- Cardinalidad reportada: 32 valores únicos
- Cardinalidad real: **6 categorías** (Belleza, Electrónica, Deportes, Hogar, Juguetería, Moda)
- Frecuencias por categoría real (aproximadas): Belleza ≈ 107.010 · Electrónica ≈ 97.188 · Deportes ≈ 106.995 · Hogar ≈ 106.590 · Juguetería ≈ 96.210 · Moda ≈ 106.043
- Inconsistencias detectadas:
  - Mayúsculas/minúsculas: *Belleza*/*belleza*/*BELLEZA*.
  - Acentos: *Electronica*/*Electrónica*/*electrónica*.
  - Espacios al inicio: *Belleza*, *Deportes*, *Moda*.
  - Sinónimo: *jugueteria* (sin tilde) además de *Juguetería*.

In [41]:
df['canal_compra'].value_counts(dropna=False)

canal_compra
Marketplace     121426
Web             121377
Tienda          121269
App             121147
tienda           15684
web              15535
app              15477
marketplace      15332
APP               7824
Marketplace       7787
 Marketplace      7780
 Tienda           7754
 Web              7748
TIENDA            7741
WEB               7737
App               7729
MARKETPLACE       7710
Web               7709
Tienda            7648
 App              7586
Name: count, dtype: int64

**Hallazgos – *canal_compra***

- Nulos: 0
- Cardinalidad reportada: 20 valores únicos
- Cardinalidad real: **4 canales** (Marketplace, Web, Tienda, App)
- Frecuencias por canal real (aproximadas): Marketplace ≈ 160.035 · Web ≈ 160.106 · Tienda ≈ 160.096 · App ≈ 159.763
- Inconsistencias detectadas:
  - Mayúsculas/minúsculas: *Web*/*WEB*/*web*, *App*/*APP*/*app*.
  - Espacios al inicio: *Marketplace*, *Tienda*, *Web*, *App*.
- Distribución muy balanceada entre los 4 canales reales.

In [42]:
df['metodo_pago'].value_counts(dropna=False)

metodo_pago
Transferencia     127962
Tarjeta           127631
PayPal            127583
Efectivo          127452
paypal             13145
efectivo           12955
transferencia      12932
tarjeta            12807
 Efectivo           6593
Tarjeta             6581
 Transferencia      6544
Transferencia       6529
 Tarjeta            6514
EFECTIVO            6513
PAYPAL              6416
Efectivo            6406
PayPal              6393
TARJETA             6385
 PayPal             6375
TRANSFERENCIA       6284
Name: count, dtype: int64

**Hallazgos – *metodo_pago***

- Nulos: 0
- Cardinalidad reportada: 20 valores únicos
- Cardinalidad real: **4 métodos** (Transferencia, Tarjeta, PayPal, Efectivo)
- Inconsistencias detectadas: mismo patrón que en *canal_compra* — mayúsculas/minúsculas (*PayPal*/*PAYPAL*/*paypal*) y espacios al inicio (*Efectivo*, *Transferencia*).
- Distribución balanceada entre los 4 métodos de pago reales.

In [43]:
df['fecha_pedido'].value_counts(dropna=False).head(20)

fecha_pedido
00/00/0000    5698
31/04/2025    5688
30/02/2025    5675
2026-13-05    5626
29/10/2025     875
14/12/2025     863
17/12/2025     861
23/12/2025     851
24/12/2025     846
31/12/2025     845
28/12/2025     845
25/10/2025     840
10/11/2025     837
28/11/2025     835
16/12/2025     834
25/11/2025     834
13/11/2025     832
19/12/2025     831
13/10/2025     828
31/10/2025     826
Name: count, dtype: int64

**Hallazgos – *fecha_pedido***

- Tipo de dato: *object* (guardada como texto, no como fecha)
- Nulos: 0
- Cardinalidad: 1.775 valores únicos
- Problemas detectados:
  1. **Formatos mezclados**: coexisten *DD/MM/YYYY* (`31/04/2025`), *D/M/YYYY* sin ceros (`19/8/2025`, `4/9/2025`) y *YYYY-MM-DD* (`2025-10-17`).
  2. **Fechas imposibles** (con altísima frecuencia, sugiriendo un error sistemático):
     - `00/00/0000`: 5.698 registros
     - `31/04/2025` (abril no tiene 31 días): 5.688 registros
     - `30/02/2025` (febrero no tiene 30 días): 5.675 registros
     - `2026-13-05` (mes 13): 5.626 registros
  3. **Fechas en el futuro**: aparecen registros con año 2027 (ej. `06/03/2027`).

In [44]:
df['fecha_actualizacion_stock'].value_counts(dropna=False).head(20)

fecha_actualizacion_stock
2026-05-07    3436
2026-05-27    3434
2026-02-17    3399
2026-06-20    3392
2026-04-04    3387
2026-06-06    3379
2026-05-14    3378
2026-04-24    3367
2026-03-19    3365
2026-03-14    3361
2026-07-04    3361
2026-06-14    3356
2026-02-19    3353
2026-04-01    3353
2026-06-28    3349
2026-04-26    3347
2026-07-12    3347
2026-05-17    3347
2026-04-19    3346
2026-03-06    3339
Name: count, dtype: int64

**Hallazgos – *fecha_actualizacion_stock***

- Tipo de dato: *object*
- Nulos: 0
- Cardinalidad: 1.177 valores únicos
- Formato: consistente (*YYYY-MM-DD*), sin ambigüedad.
- Distribución bastante uniforme entre las fechas, sin concentraciones sospechosas en el top 20.
- A diferencia de *fecha_pedido*, esta columna no presenta formatos mezclados ni fechas imposibles evidentes.

#### Hallazgos de completitud

Se auditó la completitud de las 20 columnas del dataset (640.000 registros totales).

#### Nulos reales

Seis columnas presentan valores faltantes, con un total de 316.005 celdas afectadas:

- *correo_cliente*: 105.312 registros (16.46%)
- *comentario_cliente*: 90.806 registros (14.19%)
- *nivel_satisfaccion*: 46.469 registros (7.26%)
- *ciudad_tienda*: 29.928 registros (4.68%)
- *precio_unitario*: 21.640 registros (3.38%)
- *monto_compra*: 21.640 registros (3.38%)

#### Falsos completos

**No se detectaron falsos completos** del tipo *"N/A"*, *"Sin dato"* o *"?"* en las columnas categóricas revisadas. Sin embargo, la inspección con `value_counts()` reveló un problema distinto pero muy relevante: **inconsistencia severa por mayúsculas/minúsculas, acentos, abreviaturas y espacios al inicio** en *nivel_satisfaccion*, *ciudad_tienda*, *categoria_producto*, *canal_compra* y *metodo_pago*. Este problema no afecta la completitud propiamente dicha, pero sí infla artificialmente la cardinalidad y afecta cualquier análisis basado en agrupamientos.

#### Conclusión

- La columna con mayor impacto de completitud es *correo_cliente* (16.46% de nulos), lo cual limita cualquier campaña de comunicación directa a clientes.
- *comentario_cliente* con 14.19% de nulos es esperable (no todos los clientes dejan reseña) y no debería tratarse como un problema.
- *nivel_satisfaccion* con 7.26% de nulos podría sesgar análisis futuros de satisfacción si no se trata.
- *precio_unitario* y *monto_compra* comparten los mismos 21.640 nulos, lo cual indica que probablemente son las mismas filas — habría que decidir si eliminarlas o imputarlas.
- Las columnas *nivel_lealtad* y *fecha_actualizacion_stock* son ejemplos de campos completamente limpios y consistentes.

### **2.2. Exactitud**

En esta sección se audita si los valores presentes en el dataset son **plausibles** desde el punto de vista del negocio. A diferencia de la completitud (que evalúa si el dato existe), la exactitud evalúa si el dato **tiene sentido**: un precio puede estar registrado pero ser negativo, una edad puede aparecer como 200 años, etc.

Para esto se definen **reglas de dominio**: condiciones que un dato debe cumplir para ser aceptado como válido. Cada regla se implementa como un filtro sobre `df` y se cuenta cuántas filas la incumplen con `len()`.

Se aplican **6 reglas**: 4 provenientes del enunciado y 2 propias, propuestas a partir de los hallazgos previos del perfilado.

In [45]:
# Precios negativos
filas = df[df['precio_unitario'] < 0]
print("Precios negativos:", len(filas))
filas

Precios negativos: 19629


,id_pedido,id_cliente,fecha_pedido,canal_compra,metodo_pago,ciudad_tienda,codigo_postal,categoria_producto,producto,precio_unitario,unidades_vendidas,monto_compra,edad_cliente,correo_cliente,nivel_satisfaccion,nivel_lealtad,comentario_cliente,peso_pedido_kg,fecha_actualizacion_stock,fuga_cliente
9,P622879,C155042,2025-02-23,Marketplace,efectivo,CALI,7600101,Electrónica,Audífonos inalámbricos,-124300.0,3,-372900.0,67,cliente522879@novamarket.com.co,Alto,Oro,No era lo que esperaba,15.0,2026-05-08,0
29,P507117,C105106,26/02/2026,Web,transferencia,Manizales,1700101,Electrónica,Monitor 24 pulgadas,-401600.0,3,-1204800.0,45,cliente407117@novamarket.com.co,alto,Oro,Llegó incompleto el pedido,12.7,2026-03-31,0
115,P614094,C42551,2026-02-02,App,tarjeta,Ibague,7300103,Juguetería,Carro a control remoto,-186000.0,1,-186000.0,59,cliente514094@yahoo.es,Medio,Plata,Llegó incompleto el pedido,11.5,2026-05-24,1
253,P449391,C80430,1/3/2026,web,Tarjeta,Cúcuta,800192,Juguetería,Muñeca articulada,-160100.0,1,-160100.0,26,NaN,Alto,Bronce,"Todo perfecto, llegó a tiempo",9.5,2026-03-14,0
280,P478410,C100617,13/8/2025,App,Tarjeta,Cali,7600102,BELLEZA,Perfume cítrico,-117100.0,2,-234200.0,40,cliente378410@hotmail.com,Medio,Platino,Buena atención pero demoró mucho,11.6,2026-07-23,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
639899,P461352,C148671,2025-10-23,App,Transferencia,Villavicencio,5000101,Belleza,Secador de pelo,-257900.0,1,-257900.0,39,cliente361352@novamarket.com.co,Bajo,Platino,Muy buen precio,12.9,2026-05-14,1
639912,P219430,C44597,19/07/2027,Marketplace,transferencia,CÚCUTA,5400102,Deportes,Balón de fútbol,-401500.0,1,-401500.0,56,NaN,Alto,Oro,Cumplió con lo prometido,1264.7,2026-06-11,0
639914,P244654,C40134,26/12/2025,App,Transferencia,NaN,1100107,Deportes,Colchoneta de yoga,-448400.0,1,-448400.0,23,cliente144654@novamarket.com.co,Alto,Bronce,"Excelente calidad, volvería a comprar",14.2,2026-05-08,1
639954,P343844,C35271,00/00/0000,tienda,PayPal,Cúcuta,5400101,Deportes,Pesas ajustables,-590700.0,3,-1772100.0,59,cliente243844@gmail.com,Bajo,Oro,"Excelente calidad, volvería a comprar",8.8,2026-03-04,0


#### Precios negativos

Filtré las filas donde *precio_unitario* es menor a 0. Un precio negativo no tiene sentido en un contexto de retail, así que cualquier registro que aparezca aquí es un error de captura.

**Resultado:** 19.629 registros con precio negativo.

In [46]:
# Precios mayores a 10 millones
filas = df[df['precio_unitario'] > 10000000]
print("Precios mayores a 10 millones:", len(filas))
filas

Precios mayores a 10 millones: 6164


,id_pedido,id_cliente,fecha_pedido,canal_compra,metodo_pago,ciudad_tienda,codigo_postal,categoria_producto,producto,precio_unitario,unidades_vendidas,monto_compra,edad_cliente,correo_cliente,nivel_satisfaccion,nivel_lealtad,comentario_cliente,peso_pedido_kg,fecha_actualizacion_stock,fuga_cliente
74,P620912,C57367,20/4/2025,Marketplace,Efectivo,Ibagué,7300101,DEPORTES,Colchoneta de yoga,999000000.0,1,9.990000e+08,46,NaN,Bajo,Plata,"Todo perfecto, llegó a tiempo",14.5,2026-06-02,0
123,P487008,C64904,29/01/2026,App,paypal,Cali,7600126,Hogar,Juego de sábanas,50000000.0,1,5.000000e+07,46,cliente387008@outlook.com,medio,Bronce,El domiciliario fue muy amable,11.7,2026-03-28,0
426,P216268,C137356,21/3/2025,Marketplace,PAYPAL,Villavicencio,5000101,Electrónica,Parlante bluetooth,999000000.0,3,2.997000e+09,69,NaN,Medio,Oro,No era lo que esperaba,10.8,2026-05-17,0
704,P341543,C84931,28/12/2025,Web,Efectivo,Cúcuta,5400101,Moda,Chaqueta impermeable,50000000.0,3,1.500000e+08,58,cliente241543.yahoo.es,Bajo,Bronce,El domiciliario fue muy amable,8.2,2027-08-05,0
1025,P132499,C88094,30/08/2025,Web,PayPal,NaN,1700101,Electrónica,Teclado mecánico,999000000.0,2,1.998000e+09,200,cliente32499@yahoo.es,Alto,Oro,Muy buen precio,950.6,2026-06-15,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
639441,P448387,C114179,2025-06-14,marketplace,Efectivo,Cali,7600103,JUGUETERÍA,Muñeca articulada,999000000.0,2,1.998000e+09,42,cliente348387@outlook.com,ALTO,Plata,Buena atención pero demoró mucho,13.8,2026-05-23,0
639491,P464761,C102020,2025-05-01,App,Efectivo,Barranquilla,800101,moda,Zapatos deportivos,999000000.0,0,0.000000e+00,57,cliente364761@novamarket.com.co,Alto,Platino,NaN,9.8,2026-05-24,1
639514,P207569,C155514,2025-12-26,Marketplace,Transferencia,CALI,7600102,Moda,Camisa de lino,999000000.0,1,9.990000e+08,32,cliente107569@yahoo.es,Bajo,Plata,No era lo que esperaba,10.4,2026-06-17,0
639554,P561884,C10581,2025-08-25,TIENDA,PayPal,Villavicencio,5000110,Juguetería,Carro a control remoto,50000000.0,1,5.000000e+07,34,cliente461884@outlook.com,Medio,Oro,Pésima experiencia con el empaque,11.3,2026-07-30,0


#### Precios mayores a $10.000.000

Filtré los registros con *precio_unitario* superior a 10 millones. Aunque NovaMarket vende productos de distintas categorías, un valor unitario tan alto es inusual y sugiere errores de digitación (probablemente ceros de más).

**Resultado:** 6.164 registros con precio absurdamente alto.

In [47]:
# Unidades negativas
filas = df[df['unidades_vendidas'] < 0]
print("Unidades negativas:", len(filas))
filas

Unidades negativas: 18550


,id_pedido,id_cliente,fecha_pedido,canal_compra,metodo_pago,ciudad_tienda,codigo_postal,categoria_producto,producto,precio_unitario,unidades_vendidas,monto_compra,edad_cliente,correo_cliente,nivel_satisfaccion,nivel_lealtad,comentario_cliente,peso_pedido_kg,fecha_actualizacion_stock,fuga_cliente
2,P387869,C164166,2025-03-13,Tienda,PayPal,Cúcuta,5400101,Juguetería,Rompecabezas 1000 piezas,597600.0,-3,-1792800.0,32,cliente287869@novamarket.com.co,Bajo,Oro,El domiciliario fue muy amable,14.3,2026-02-26,0
7,P583062,C94873,19/08/2027,TIENDA,PayPal,Medellín,500103,Deportes,Bicicleta estática,190900.0,-1,-190900.0,26,cliente483062@outlook.com,Alto,Platino,NaN,8.4,2024-03-14,0
23,P517960,C145474,01/03/2026,Web,Transferencia,Bogotá,5400103,moda,Chaqueta impermeable,326800.0,-1,-326800.0,62,cliente417960@novamarket.com.co,Alto,Oro,"Excelente calidad, volvería a comprar",13.5,2026-04-10,0
71,P295334,C105354,2025-02-12,Web,PayPal,Villavicencio,5000110,MODA,Camisa de lino,202900.0,-3,-608700.0,65,cliente195334@outlook.com,Alto,Platino,Buena atención pero demoró mucho,12.9,2026-03-08,0
94,P371313,C86387,09/02/2026,marketplace,Tarjeta,cucuta,5400102,Belleza,Perfume cítrico,556200.0,-1,-556200.0,33,cliente271313@novamarket.com.co,Alto,Platino,El domiciliario fue muy amable,12.3,2026-08-01,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
639762,P240670,C113313,19/04/2026,Marketplace,PayPal,MANIZALES,1700101,Moda,Chaqueta impermeable,254600.0,-1,-254600.0,27,cliente140670.outlook.com,MEDIO,Plata,"Excelente calidad, volvería a comprar",13.9,2026-02-27,0
639795,P358108,C52776,2026-01-23,App,Efectivo,Bucaramanga,6800101,Moda,Zapatos deportivos,467800.0,-3,-1403400.0,35,cliente258108@outlook.com,Bajo,Platino,Llegó incompleto el pedido,14.2,2026-03-17,0
639933,P313641,C100317,14/07/2025,Marketplace,Tarjeta,Manizales,1700101,Moda,Chaqueta impermeable,434200.0,-3,-1302600.0,63,cliente213641@hotmail.com,Medio,Bronce,Llegó incompleto el pedido,10.7,2026-05-03,0
639945,P192825,C77013,31/08/2025,Tienda,PayPal,Bucaramanga,6800104,belleza,Kit de brochas,203500.0,-3,-610500.0,31,cliente92825@yahoo.es,Medio,Oro,"Todo perfecto, llegó a tiempo",8.4,2023-09-15,0


#### Unidades negativas

Filtré las filas donde *unidades_vendidas* es menor a 0. No es posible vender una cantidad negativa de un producto, así que estos casos también corresponden a errores de captura.

**Resultado:** 18.550 registros con unidades negativas.

In [48]:
# Edad fuera de 0 a 120
filas = df[(df['edad_cliente'] < 0) | (df['edad_cliente'] > 120)]
print("Edades fuera de rango:", len(filas))
filas

Edades fuera de rango: 27829


,id_pedido,id_cliente,fecha_pedido,canal_compra,metodo_pago,ciudad_tienda,codigo_postal,categoria_producto,producto,precio_unitario,unidades_vendidas,monto_compra,edad_cliente,correo_cliente,nivel_satisfaccion,nivel_lealtad,comentario_cliente,peso_pedido_kg,fecha_actualizacion_stock,fuga_cliente
62,P181570,C122603,06/01/2025,App,Tarjeta,Barranquilla,1700101,hogar,Lámpara de escritorio,379400.0,1,379400.0,200,cliente81570@gmail.com,ALTO,Oro,"Todo perfecto, llegó a tiempo",12.4,2026-06-19,0
72,P366245,C127008,2025-02-05,Tienda,PayPal,CALI,7600105,belleza,Perfume cítrico,345300.0,2,690600.0,150,cliente266245@hotmail.com,Bajo,Bronce,Me costó encontrar el producto en la app,8.8,2026-07-23,0
83,P266054,C25619,2025-10-21,Tienda,Tarjeta,Pereira,6600101,Belleza,Secador de pelo,581300.0,2,1162600.0,200,cliente166054@hotmail.com,NaN,Platino,Rápido y sin problemas,12.9,2026-05-18,0
97,P360672,C34925,4/5/2025,Web,Transferencia,BARRANQUILLA,800108,Belleza,Perfume cítrico,329000.0,2,658000.0,200,cliente260672@gmail.com,Bajo,Oro,Rápido y sin problemas,14.4,2026-04-21,0
150,P160210,C33233,17/01/2026,Tienda,Efectivo,Bogotá,1100111,Belleza,Secador de pelo,471200.0,1,471200.0,-5,cliente60210@gmail.com,Alto,Bronce,"Excelente calidad, volvería a comprar",9.6,2026-06-11,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
639948,P592410,C65292,16/10/2025,App,PayPal,Medellín,500104,juguetería,Carro a control remoto,564500.0,1,564500.0,-5,cliente492410@hotmail.com,Alto,Plata,No era lo que esperaba,10.8,2026-03-27,0
639953,P131668,C68615,2026-05-08,App,Efectivo,Cúcuta,5400102,electronica,Monitor 24 pulgadas,407300.0,-1,-407300.0,200,cliente31668.hotmail.com,Medio,Plata,"Excelente calidad, volvería a comprar",13.7,2026-02-16,0
639980,P630605,C33803,3/8/2025,App,paypal,BARRANQUILLA,800109,moda,Camisa de lino,245700.0,1,245700.0,200,cliente530605@novamarket.com.co,Medio,Bronce,Muy buen precio,8.8,2026-03-01,0
639982,P231923,C27475,20/01/2025,Marketplace,Tarjeta,ibague,7300101,Belleza,Secador de pelo,213800.0,5,1069000.0,200,NaN,Bajo,Bronce,Llegó incompleto el pedido,12.3,2027-08-11,1


#### Edades fuera del rango [0, 120]

Filtré los registros con *edad_cliente* menor a 0 o mayor a 120. Usé el operador `|` porque basta con que se cumpla una de las dos condiciones para que la edad sea implausible.

**Resultado:** 27.829 registros con edad fuera de rango, incluyendo valores negativos y edades superiores al límite biológico razonable.

In [49]:
# Unidades mayores a 100
filas = df[df['unidades_vendidas'] > 100]
print("Unidades mayores a 100:", len(filas))
filas

Unidades mayores a 100: 18526


,id_pedido,id_cliente,fecha_pedido,canal_compra,metodo_pago,ciudad_tienda,codigo_postal,categoria_producto,producto,precio_unitario,unidades_vendidas,monto_compra,edad_cliente,correo_cliente,nivel_satisfaccion,nivel_lealtad,comentario_cliente,peso_pedido_kg,fecha_actualizacion_stock,fuga_cliente
42,P343224,C30024,31/3/2026,Tienda,Transferencia,IBAGUÉ,1100102,BELLEZA,Secador de pelo,114400.0,5000,5.720000e+08,41,cliente243224@gmail.com,Medio,Platino,Muy buen precio,14.3,2026-04-26,0
45,P182089,C108998,10/11/2025,Web,Efectivo,NaN,6600107,Hogar,Lámpara de escritorio,263300.0,5000,1.316500e+09,63,cliente82089@outlook.com,medio,Platino,El producto llegó dañado,1103.2,2026-02-27,0
93,P559710,C163756,27/2/2026,App,tarjeta,Ibagué,7300101,Electrónica,Teclado mecánico,512300.0,9999,5.122488e+09,48,cliente459710@hotmail.com,Alto,Oro,Rápido y sin problemas,9.1,2026-07-31,0
99,P317186,C85461,24/10/2025,Web,Tarjeta,NaN,6800112,Deportes,Bicicleta estática,688200.0,5000,3.441000e+09,24,NaN,Bajo,Bronce,Muy buen precio,14.2,2026-03-07,1
105,P261739,C92152,14/7/2025,Web,Efectivo,Cali,1100104,Belleza,Kit de brochas,90900.0,9999,9.089091e+08,64,cliente161739@gmail.com,Alto,Bronce,Buena atención pero demoró mucho,10.5,2026-03-26,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
639794,P350280,C60965,14/1/2026,marketplace,Efectivo,Villavicencio,5000101,Juguetería,Carro a control remoto,NaN,9999,NaN,46,cliente250280@gmail.com,medio,Bronce,NaN,13.5,2026-06-17,0
639842,P653257,C140209,11/1/2026,Marketplace,Efectivo,Villavicencio,5000109,Deportes,Balón de fútbol,795900.0,5000,3.979500e+09,69,cliente553257@yahoo.es,Medio,Oro,Me costó encontrar el producto en la app,8.6,2026-07-27,0
639886,P351993,C117339,2026-01-26,Marketplace,Tarjeta,PEREIRA,6600102,Belleza,Perfume cítrico,56800.0,5000,2.840000e+08,71,NaN,Alto,Bronce,Muy buen precio,8.4,2023-11-23,0
639901,P636304,C89459,24/2/2025,App,tarjeta,Cúcuta,5400101,Juguetería,Muñeca articulada,329300.0,5000,1.646500e+09,55,cliente536304@gmail.com,NaN,Bronce,NaN,11.4,2026-07-16,0


#### Unidades mayores a 100 (regla propia)

Además de las reglas del enunciado, propuse esta regla adicional. En un dataset de retail dirigido a consumidor final, un pedido de más de 100 unidades del mismo producto es muy poco frecuente y probablemente indica un error de captura (por ejemplo, valores como 9999 que aparecen como máximo en el describe de *unidades_vendidas*).

**Resultado:** 18.526 registros con más de 100 unidades por pedido.

In [50]:
# Correos sin @
correos_no_nulos = df[df['correo_cliente'].notna()]
filas = correos_no_nulos[~correos_no_nulos['correo_cliente'].str.contains('@')]
print("Correos sin @:", len(filas))
filas

Correos sin @: 51532


,id_pedido,id_cliente,fecha_pedido,canal_compra,metodo_pago,ciudad_tienda,codigo_postal,categoria_producto,producto,precio_unitario,unidades_vendidas,monto_compra,edad_cliente,correo_cliente,nivel_satisfaccion,nivel_lealtad,comentario_cliente,peso_pedido_kg,fecha_actualizacion_stock,fuga_cliente
18,P693809,C155675,19/2/2026,App,Efectivo,Villavicencio,5000103,Juguetería,Muñeca articulada,316200.0,1,316200.0,56,cliente593809.hotmail.com,bajo,Oro,"Excelente calidad, volvería a comprar",11.6,2023-10-03,0
25,P683717,C114516,06/10/2025,web,tarjeta,Barranquilla,800101,Electrónica,Teclado mecánico,305800.0,1,305800.0,65,cliente583717.outlook.com,Bajo,Platino,"Todo perfecto, llegó a tiempo",10.6,2026-02-26,0
36,P332336,C46793,15/01/2026,app,Tarjeta,Bogota,1100119,Deportes,Bicicleta estática,158700.0,3,476100.0,49,cliente232336.gmail.com,Medio,Bronce,"Excelente calidad, volvería a comprar",9.0,2026-05-03,0
96,P284801,C38220,08/07/2025,Tienda,Transferencia,Cartagena de Indias,1300111,electrónica,Audífonos inalámbricos,280200.0,1,280200.0,23,cliente184801.gmail.com,Medio,Platino,NaN,14.8,2026-04-05,0
103,P527463,C77007,4/1/2026,Tienda,PayPal,Pereira,6600101,JUGUETERÍA,Carro a control remoto,79400.0,5,397000.0,55,cliente427463.hotmail.com,Medio,Bronce,No era lo que esperaba,9.7,2026-06-10,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
639955,P588879,C146577,2025-08-28,web,Efectivo,manizales,1700111,Juguetería,Rompecabezas 1000 piezas,455700.0,1,455700.0,75,cliente488879.hotmail.com,Alto,Bronce,El domiciliario fue muy amable,13.4,2026-05-11,0
639963,P608802,C54387,24/7/2025,Marketplace,Tarjeta,Santa Marta,4700102,DEPORTES,Colchoneta de yoga,501700.0,3,1505100.0,60,cliente508802.outlook.com,Alto,Oro,Cumplió con lo prometido,9.0,2026-02-22,0
639965,P308655,C31167,2025-08-02,Marketplace,Transferencia,Ibagué,7300101,Electrónica,Parlante bluetooth,214100.0,3,642300.0,49,cliente208655.gmail.com,Medio,Oro,Llegó incompleto el pedido,11.7,2026-03-28,0
639972,P321630,C87019,23/02/2025,TIENDA,PayPal,cali,7600117,jugueteria,Set de bloques,246800.0,2,493600.0,65,cliente221630.outlook.com,Alto,Platino,Rápido y sin problemas,13.7,2026-03-13,0


#### Correos sin `@` (regla propia)

Como segunda regla propia, revisé el formato de la columna *correo_cliente*. Un correo válido siempre debe contener el símbolo `@`. Hice el filtro en dos pasos:

1. Primero me quedé con los correos que no son nulos, para no mezclar este hallazgo con el de completitud.
2. Sobre esos, filtré los que no contienen `@`.

**Resultado:** 51.532 registros con correo mal formateado (existe la cadena de texto pero no cumple la estructura mínima de un correo electrónico).

#### Hallazgos de exactitud

Se aplicaron 6 reglas de dominio al dataset de NovaMarket. Los resultados fueron:

| # | Regla | Registros afectados |
|---|---|---|
| 1 | *precio_unitario* < 0 | 19.629 |
| 2 | *precio_unitario* > $10.000.000 | 6.164 |
| 3 | *unidades_vendidas* < 0 | 18.550 |
| 4 | *edad_cliente* fuera de [0, 120] | 27.829 |
| 5 | *unidades_vendidas* > 100 (regla propia) | 18.526 |
| 6 | *correo_cliente* sin `@` (regla propia) | 51.532 |

**Total de registros con al menos un problema de exactitud: hasta 142.230**, aunque algunas filas pueden acumular varios errores simultáneamente (por ejemplo, un mismo pedido con precio negativo y edad fuera de rango).

#### Conclusiones

- Las reglas 1, 3 y 4 evidencian **errores de captura básicos**: valores negativos o extremos en campos que nunca deberían tenerlos. En conjunto afectan cerca del 10% del dataset.
- La regla 2 confirma la presencia de **outliers extremos en precio_unitario**, con valores absurdamente altos que deben tratarse antes de cualquier análisis estadístico.
- La regla 5 refuerza el hallazgo del describe de *unidades_vendidas*: valores como 9999 no corresponden a compras reales.
- La regla 6 revela un problema de **integridad estructural** en *correo_cliente* que impacta directamente cualquier campaña de comunicación con clientes: más de 51 mil correos existen pero no son válidos como dirección de correo.
- Combinando completitud y exactitud, se estima que **aproximadamente el 22% del dataset presenta al menos un problema de calidad** (nulos o valores implausibles), lo cual justifica una fase de limpieza sistemática antes de cualquier análisis posterior.

### **2.3. Consistencia**

#### Ciudad

In [38]:
df["ciudad_tienda"].value_counts().sort_index()

ciudad_tienda
 Barranquilla    3260
 Bogotá          2662
 Bucaramanga     3091
 Cali            3205
 Cartagena       3173
                 ... 
medellin         6347
pereira          7816
santa marta      3150
santamarta       3202
villavicencio    7708
Name: count, Length: 88, dtype: int64

In [33]:
df['ciudad_tienda'].unique()

array(['bogota', 'MZL', 'Cúcuta', 'Cucuta', 'Bucaramanga', 'MDE',
       'pereira', 'Medellín ', 'MANIZALES', 'CALI', 'Bogota',
       'Santa Marta', ' Ibagué', 'Villavicencio', 'B/manga', 'Pereira',
       'CTG', 'Bogotá', 'Cali', 'Barranquilla', 'BAQ', 'Cúcuta ',
       'B/quilla', ' Manizales', ' Barranquilla', 'Ibagué ', nan,
       'IBAGUE', 'cartagena', 'IBAGUÉ', 'SantaMarta', 'Medellín',
       'Ibagué', 'Cartagena', 'CUCUTA', 'Manizales', 'Medellin',
       'SANTA MARTA', 'Bucaramanga ', 'cali', 'Ibague', 'VILLAVICENCIO',
       ' Cali', ' Cúcuta', 'MEDELLÍN', 'Cali ', ' Cartagena', 'cucuta',
       'Cartagena de Indias', 'BARRANQUILLA', 'PEREIRA', 'Villavicencio ',
       'CARTAGENA', 'Sta. Marta', 'BOG', 'Medellin, Ant.', 'Manizales ',
       ' Villavicencio', 'Bogota D.C.', 'manizales', 'CALI ',
       'BUCARAMANGA', 'Bogotá ', 'santamarta', 'BGA', 'medellin',
       ' Bogotá', 'BOGOTA', 'villavicencio', 'barranquilla', 'SMR',
       'bucaramanga', 'Santiago de Cali', 'Perei

In [36]:
conteos = df['ciudad_tienda'].value_counts(dropna=False)

ciudades = {
    'Bogotá': ['Bogotá', 'Bogotá ', ' Bogotá', 'Bogota', 'bogota', 'BOGOTA',
               'BOGOTÁ', 'Bogota D.C.', 'bogota d.c.', 'BOGOTA DC', 'BOG'],

    'Medellín': ['Medellín', 'Medellín ', ' Medellín', 'Medellin',
                 'MEDELLIN', 'MEDELLÍN', 'medellin', 'MDE', 'Medellin, Ant.'],

    'Cali': ['Cali', 'CALI', 'CALI ', ' Cali', 'Cali ', 'cali',
             'Santiago de Cali'],

    'Barranquilla': ['Barranquilla', ' Barranquilla', 'Barranquilla ',
                     'BARRANQUILLA', 'barranquilla', 'BAQ', 'B/quilla'],

    'Bucaramanga': ['Bucaramanga', 'Bucaramanga ', ' Bucaramanga',
                    'BUCARAMANGA', 'bucaramanga', 'BGA', 'B/manga'],

    'Cartagena': ['Cartagena', 'Cartagena ', ' Cartagena', 'cartagena',
                  'CARTAGENA', 'CTG', 'Cartagena de Indias'],

    'Cúcuta': ['Cúcuta', 'Cúcuta ', ' Cúcuta', 'Cucuta',
               'CUCUTA', 'CÚCUTA', 'cucuta'],

    'Ibagué': ['Ibagué', 'Ibagué ', ' Ibagué', 'Ibague',
               'IBAGUE', 'IBAGUÉ', 'ibague'],

    'Manizales': ['Manizales', 'Manizales ', ' Manizales',
                  'MANIZALES', 'manizales', 'MZL',
                  'Manizales, Caldas'],

    'Pereira': ['Pereira', 'Pereira ', ' Pereira',
                'pereira', 'PEREIRA'],

    'Santa Marta': ['Santa Marta', 'Santa Marta ', ' Santa Marta',
                    'SANTA MARTA', 'santa marta', 'santamarta',
                    'SantaMarta', 'Sta. Marta', 'SMR'],

    'Villavicencio': ['Villavicencio', 'Villavicencio ',
                      ' Villavicencio', 'VILLAVICENCIO',
                      'villavicencio']
}

for ciudad, variantes in ciudades.items():
    total = conteos.reindex(variantes, fill_value=0).sum()
    print(f"{ciudad}: {total}")

Bogotá: 50810
Medellín: 50775
Cali: 51002
Barranquilla: 50884
Bucaramanga: 50840
Cartagena: 51079
Cúcuta: 50735
Ibagué: 50636
Manizales: 50991
Pereira: 50863
Santa Marta: 50755
Villavicencio: 50702


##### Hallazgos de consistencia para ciudad

Se identificaron inconsistencias en la variable ciudad_tienda, principalmente relacionadas con la forma en que se registran los nombres de las ciudades. Se encontraron diferencias de mayúsculas y minúsculas, espacios en blanco, tildes, abreviaciones y nombres alternativos para una misma ciudad.

Por ejemplo:

Bogotá: Bogota, Bogotá, BOG, BOGOTA DC, Bogota D.C., bogota d.c., entre otras.
Medellín: Medellin, Medellín, MEDELLIN, MDE, Medellin, Ant..
Bucaramanga: Bucaramanga, B/manga, BGA, BUCARAMANGA.
Barranquilla: Barranquilla, BAQ, B/quilla, BARRANQUILLA.
Cartagena: Cartagena, CTG, CARTAGENA, Cartagena de Indias.
Santa Marta: Santa Marta, SantaMarta, SMR, Sta. Marta, santa marta.
Cali: Cali, CALI, Santiago de Cali, además de variantes con espacios.
Cúcuta: Cúcuta, Cucuta, CUCUTA, CÚCUTA, incluyendo registros con espacios.

También se observan espacios al inicio o al final de algunos registros, por ejemplo Medellín, Cali o Bucaramanga , y se identifica la presencia de valores nulos (NaN).

Estas inconsistencias pueden provocar que una misma ciudad sea contabilizada como categorías diferentes. Por ejemplo, medellin, Medellín, MEDELLIN y MDE aparecen como registros independientes, aunque representan la misma ciudad. Esto puede distorsionar las frecuencias, agrupaciones y visualizaciones realizadas sobre esta variable.

Conclusión: se recomienda realizar una estandarización de ciudad_tienda, eliminando espacios innecesarios, unificando mayúsculas/minúsculas y tildes, y estableciendo un catálogo único de ciudades para homologar las abreviaciones y nombres alternativos antes de realizar análisis agrupados.

#### Hallazgos de consistencia para ciudad:

La variable `ciudad_tienda` presenta inconsistencias en la estandarización de los nombres de las ciudades. De los **620.000 registros analizados, 4.7 % presentan valores nulos**. En los registros no nulos se identificaron múltiples variantes para una misma ciudad, debido principalmente al uso de mayúsculas y minúsculas, tildes, espacios adicionales, abreviaciones y nombres alternativos.

Por ejemplo:

Bogotá: Bogota, Bogotá, BOG, BOGOTA DC, Bogota D.C., bogota d.c., entre otras.
Medellín: Medellin, Medellín, MEDELLIN, MDE, Medellin, Ant..
Bucaramanga: Bucaramanga, B/manga, BGA, BUCARAMANGA.
Barranquilla: Barranquilla, BAQ, B/quilla, BARRANQUILLA.
Cartagena: Cartagena, CTG, CARTAGENA, Cartagena de Indias.
Santa Marta: Santa Marta, SantaMarta, SMR, Sta. Marta, santa marta.
Cali: Cali, CALI, Santiago de Cali, además de variantes con espacios.
Cúcuta: Cúcuta, Cucuta, CUCUTA, CÚCUTA, incluyendo registros con espacios.

Estas inconsistencias pueden provocar que una misma ciudad sea contabilizada como categorías diferentes. Por ejemplo, medellin, Medellín, MEDELLIN y MDE aparecen como registros independientes, aunque representan la misma ciudad. Esto puede distorsionar las frecuencias, agrupaciones y visualizaciones realizadas sobre esta variable.

También podemos ver que si agrupamos correctamente los datos el conteo de cada ciudad sube en miles de registros.

#### Categoría de producto

In [29]:
df["categoria_producto"].value_counts()

categoria_producto
Belleza         56378
Electrónica     56261
Deportes        56019
Hogar           55846
Juguetería      55499
Moda            55300
moda            20354
hogar           20352
deportes        20351
belleza         20273
electronica     10433
JUGUETERÍA      10332
HOGAR           10290
Electrónica     10267
DEPORTES        10264
Deportes        10214
Belleza         10195
Moda            10182
 Juguetería     10170
Hogar           10157
 Deportes       10147
electrónica     10142
 Moda           10136
juguetería      10133
BELLEZA         10114
ELECTRÓNICA     10085
 Belleza        10052
Juguetería      10047
jugueteria      10029
MODA            10025
 Electrónica    10008
 Hogar           9945
Name: count, dtype: int64

In [37]:
df['categoria_producto'].unique()

array(['Moda ', 'Electrónica', 'Juguetería', 'deportes', 'Hogar ',
       'Deportes', 'Moda', ' Deportes', 'moda', 'electronica',
       ' Juguetería', 'jugueteria', 'Electrónica ', 'Belleza ',
       ' Belleza', 'Hogar', 'DEPORTES', 'hogar', 'BELLEZA', 'Belleza',
       'Juguetería ', 'JUGUETERÍA', 'electrónica', ' Hogar', 'MODA',
       'belleza', 'HOGAR', 'juguetería', ' Electrónica', 'ELECTRÓNICA',
       'Deportes ', ' Moda'], dtype=object)

In [39]:
conteos = df['categoria_producto'].value_counts(dropna=False)

categorias = {
    'Moda': [
        'Moda', 'Moda ', ' Moda', 'moda', 'MODA'
    ],

    'Electrónica': [
        'Electrónica', 'Electrónica ', ' Electrónica',
        'electrónica', 'electronica', 'ELECTRÓNICA'
    ],

    'Juguetería': [
        'Juguetería', 'Juguetería ', ' Juguetería',
        'juguetería', 'jugueteria', 'JUGUETERÍA'
    ],

    'Deportes': [
        'Deportes', 'Deportes ', ' Deportes',
        'deportes', 'DEPORTES'
    ],

    'Hogar': [
        'Hogar', 'Hogar ', ' Hogar',
        'hogar', 'HOGAR'
    ],

    'Belleza': [
        'Belleza', 'Belleza ', ' Belleza',
        'belleza', 'BELLEZA'
    ]
}

for categoria, variantes in categorias.items():
    total = conteos.reindex(variantes, fill_value=0).sum()
    print(f"{categoria}: {total}")

Moda: 105997
Electrónica: 107196
Juguetería: 106210
Deportes: 106995
Hogar: 106590
Belleza: 107012


##### Hallazgos de consistencia para categoría producto

La variable `categoria_producto` contiene 620.000 registros y no presenta valores nulos, por lo que existe una cobertura completa de información en esta variable. Sin embargo, se identificaron inconsistencias en la forma de registrar las categorías de producto.

Actualmente se encuentran **32 categorías diferentes**, aunque estas corresponden realmente a solo **6 categorías principales**: Belleza, Electrónica, Deportes, Hogar, Juguetería y Moda. Las diferencias se deben principalmente al uso inconsistente de **mayúsculas y minúsculas, espacios al inicio o al final, y variaciones en el uso de tildes**.

Por ejemplo, la categoría **Moda** aparece como `Moda`, `moda`, `MODA`, ` Moda` y `Moda `. De manera similar, **Electrónica** presenta variantes como `Electrónica`, `electrónica`, `ELECTRÓNICA`, ` Electrónica` y `Electrónica `. También se encontraron casos como `jugueteria` y `Juguetería`, donde además existe una diferencia en el uso de la tilde.

Estas inconsistencias afectan la calidad del análisis debido a que registros pertenecientes a una misma categoría pueden ser contabilizados como categorías independientes. Por ejemplo, al realizar un `value_counts()` sin una estandarización previa, `Moda` y `moda` aparecen como categorías diferentes, aunque representan el mismo tipo de producto.

En consecuencia, aunque el **100 % de los registros contiene información**, la variable presenta un problema de **consistencia categórica**. Se recomienda estandarizar los valores mediante la eliminación de espacios innecesarios y la homologación de mayúsculas, minúsculas y tildes, reduciendo las 32 categorías observadas a las **6 categorías reales** antes de realizar análisis estadísticos o visualizaciones.


#### Hallazgos de consistencia para categoría de producto

La variable `categoria_producto` contiene 620.000 registros y no presenta valores nulos, por lo que existe una cobertura completa de información en esta variable. Sin embargo, se identificaron inconsistencias en la forma de registrar las categorías de producto.

Actualmente se encuentran **32 categorías diferentes**, aunque estas corresponden realmente a solo **6 categorías principales**: Belleza, Electrónica, Deportes, Hogar, Juguetería y Moda. Las diferencias se deben principalmente al uso inconsistente de mayúsculas y minúsculas, espacios al inicio o al final, y variaciones en el uso de tildes**.

Por ejemplo, la categoría Moda aparece como `Moda`, `moda`, `MODA`, ` Moda` y `Moda `. De manera similar, Electrónica presenta variantes como `Electrónica`, `electrónica`, `ELECTRÓNICA`, ` Electrónica` y `Electrónica `. También se encontraron casos como `jugueteria` y `Juguetería`, donde además existe una diferencia en el uso de la tilde.

Estas inconsistencias afectan la calidad del análisis debido a que registros pertenecientes a una misma categoría pueden ser contabilizados como categorías independientes. Por ejemplo, al realizar un `value_counts()` sin una estandarización previa, `Moda` y `moda` aparecen como categorías diferentes, aunque representan el mismo tipo de producto.

En consecuencia, aunque el 100 % de los registros contiene informació*, la variable presenta un problema de **consistencia categórica**. Se recomienda estandarizar los valores mediante la eliminación de espacios innecesarios y la homologación de mayúsculas, minúsculas y tildes, reduciendo las 32 categorías observadas a las **6 categorías reales** antes de realizar análisis estadísticos o visualizaciones.




#### Canal_compra

In [18]:
df["canal_compra"].value_counts()

canal_compra
Marketplace     121426
Web             121377
Tienda          121269
App             121147
tienda           15684
web              15535
app              15477
marketplace      15332
APP               7824
Marketplace       7787
 Marketplace      7780
 Tienda           7754
 Web              7748
TIENDA            7741
WEB               7737
App               7729
MARKETPLACE       7710
Web               7709
Tienda            7648
 App              7586
Name: count, dtype: int64

In [ ]:
df['canal_compra'].unique()

In [25]:
conteos = df['canal_compra'].value_counts(dropna=False)

canales = {
    'Marketplace': [
        'Marketplace',
        'marketplace',
        'MARKETPLACE',
        'Marketplace ',
        ' Marketplace'
    ],

    'Web': [
        'Web',
        'web',
        'WEB',
        'Web ',
        ' Web'
    ],

    'Tienda': [
        'Tienda',
        'tienda',
        'TIENDA',
        'Tienda ',
        ' Tienda'
    ],

    'App': [
        'App',
        'app',
        'APP',
        'App ',
        ' App'
    ]
}

for canal, variantes in canales.items():
    total = conteos.reindex(variantes, fill_value=0).sum()
    print(f"{canal}: {total}")

Marketplace: 160035
Web: 160106
Tienda: 160096
App: 159763


##### Hallazgos de consistencia para canal de compra

La variable `canal_compra` presenta problemas de consistencia en el formato de sus valores, aunque no presenta problemas de completitud. Se identificaron diferentes formas de escritura para los canales de compra debido principalmente al uso de mayúsculas, minúsculas y espacios adicionales.

A partir de la homologación de las variantes, se identificaron **4 canales de compra principales**: Marketplace, Web, Tienda y App. La distribución consolidada es la siguiente:

| Canal de compra |   Registros | Porcentaje |
| --------------- | ----------: | ---------: |
| Marketplace     |     160.035 |    25,01 % |
| Web             |     160.106 |    25,02 % |
| Tienda          |     160.096 |    25,01 % |
| App             |     159.763 |    24,96 % |
| **Total**       | **640.000** |  **100 %** |

La distribución entre los cuatro canales es bastante equilibrada, ya que cada uno representa aproximadamente el 25 % de los registros. La diferencia entre el canal con mayor cantidad de registros (`Web`, con 160.106) y el de menor cantidad (`App`, con 159.763) es de únicamente 343 registros

No se encontraron valores nulos en `canal_compra`, por lo que la variable presenta una **completitud del 100 %**. Sin embargo, antes de la homologación existían múltiples variantes para representar un mismo canal, por ejemplo `Marketplace`, `marketplace`, `MARKETPLACE` y valores con espacios adicionales. Estas diferencias pueden provocar que los canales sean contabilizados como categorías independientes y afectar los análisis de frecuencia y visualizaciones.

En conclusión, el principal hallazgo corresponde a un **problema de consistencia categórica**, no de completitud. Se recomienda mantener una única representación estandarizada para cada uno de los cuatro canales antes de realizar análisis posteriores.


#### Método de pago

In [36]:
df["metodo_pago"].value_counts()

metodo_pago
Transferencia     127962
Tarjeta           127631
PayPal            127583
Efectivo          127452
paypal             13145
efectivo           12955
transferencia      12932
tarjeta            12807
 Efectivo           6593
Tarjeta             6581
 Transferencia      6544
Transferencia       6529
 Tarjeta            6514
EFECTIVO            6513
PAYPAL              6416
Efectivo            6406
PayPal              6393
TARJETA             6385
 PayPal             6375
TRANSFERENCIA       6284
Name: count, dtype: int64

In [35]:
df['metodo_pago'].unique()

array(['Efectivo', ' Efectivo', 'PayPal', 'transferencia', 'Tarjeta',
       'efectivo', 'Transferencia ', 'Transferencia', 'PayPal ',
       'tarjeta', 'Tarjeta ', ' PayPal', 'TRANSFERENCIA', 'PAYPAL',
       'paypal', 'EFECTIVO', ' Tarjeta', 'Efectivo ', ' Transferencia',
       'TARJETA'], dtype=object)

In [34]:
conteos = df['metodo_pago'].value_counts(dropna=False)

metodos_pago = {
    'Efectivo': [
        'Efectivo',
        'efectivo',
        'EFECTIVO',
        ' Efectivo',
        'Efectivo '
    ],

    'PayPal': [
        'PayPal',
        'paypal',
        'PAYPAL',
        ' PayPal',
        'PayPal '
    ],

    'Transferencia': [
        'Transferencia',
        'transferencia',
        'TRANSFERENCIA',
        ' Transferencia',
        'Transferencia '
    ],

    'Tarjeta': [
        'Tarjeta',
        'tarjeta',
        'TARJETA',
        ' Tarjeta',
        'Tarjeta '
    ]
}

for metodo, variantes in metodos_pago.items():
    total = conteos.reindex(variantes, fill_value=0).sum()
    print(f"{metodo}: {total}")

Efectivo: 159919
PayPal: 159912
Transferencia: 160251
Tarjeta: 159918


##### Hallazgos de consistencia método de pago

Se analizó la consistencia de la columna `metodo_pago`. La variable presenta un total de **20 variantes de formato distintas** para representar únicamente 4 métodos de pago válidos (*Efectivo*, *PayPal*, *Transferencia*, *Tarjeta*), además de un volumen de registros nulos.

Se detectaron las siguientes inconsistencias:

1. **Espacios en blanco en los extremos (*Leading/Trailing spaces*):** Presencia de espacios al inicio (ej. `' Efectivo'`) o al final (ej. `'Transferencia '`), generando categorías duplicadas.
2. **Inconsistencia en mayúsculas y minúsculas:**
   * Formato minúscula: `efectivo`, `paypal`, `transferencia`, `tarjeta`
   * Formato mayúscula sostenida: `EFECTIVO`, `PAYPAL`, `TRANSFERENCIA`, `TARJETA`
   * Formato título: `Efectivo`, `PayPal`, `Transferencia`, `Tarjeta`
3. **Valores nulos (`NaN`):** Se registraron **29,928** valores faltantes, correspondientes al **4.47%** del total de la muestra (669,928 registros).

Mientras que la proporción de las variables quedó así:

| Método de Pago Unificado | Frecuencia Absoluta | Proporción (%) |
| :--- | :---: | :---: |
| **Transferencia** | 160,251 | 23.92% |
| **Efectivo** | 159,919 | 23.87% |
| **Tarjeta** | 159,918 | 23.87% |
| **PayPal** | 159,912 | 23.87% |
| **Nulos (`NaN`)** | 29,928 | 4.47% |
| **Total** | **669,928** | **100.00%** |


#### Nivel de satisfacción

In [24]:
df["metodo_pagotisfaccion"].value_counts()

nivel_satisfaccion
Alto      166739
Medio     165777
Bajo      165309
alto       12803
medio      12661
bajo       12585
BAJO        6582
 Alto       6554
 Medio      6444
Medio       6429
 Bajo       6383
MEDIO       6347
Alto        6335
ALTO        6315
Bajo        6268
Name: count, dtype: int64

In [22]:
df['nivel_satisfaccion'].unique()

array(['Bajo', 'Medio', ' Bajo', 'Alto', 'ALTO', 'Alto ', 'medio',
       ' Medio', 'bajo', 'alto', ' Alto', nan, 'Bajo ', 'MEDIO', 'BAJO',
       'Medio '], dtype=object)

#### Nivel de satisfacción

In [33]:
df["nivel_satisfaccion"].value_counts()

nivel_satisfaccion
Alto      166739
Medio     165777
Bajo      165309
alto       12803
medio      12661
bajo       12585
BAJO        6582
 Alto       6554
 Medio      6444
Medio       6429
 Bajo       6383
MEDIO       6347
Alto        6335
ALTO        6315
Bajo        6268
Name: count, dtype: int64

In [32]:
df['nivel_satisfaccion'].unique()

array(['Bajo', 'Medio', ' Bajo', 'Alto', 'ALTO', 'Alto ', 'medio',
       ' Medio', 'bajo', 'alto', ' Alto', nan, 'Bajo ', 'MEDIO', 'BAJO',
       'Medio '], dtype=object)

In [31]:
conteos = df['nivel_satisfaccion'].value_counts(dropna=False)

niveles_satisfaccion = {
    'Alto': [
        'Alto',
        'alto',
        'ALTO',
        ' Alto',
        'Alto '
    ],

    'Medio': [
        'Medio',
        'medio',
        'MEDIO',
        ' Medio',
        'Medio '
    ],

    'Bajo': [
        'Bajo',
        'bajo',
        'BAJO',
        ' Bajo',
        'Bajo '
    ]
}

for niveles, variantes in niveles_satisfaccion.items():
    total = conteos.reindex(variantes, fill_value=0).sum()
    print(f"{niveles}: {total}")

Alto: 198746
Medio: 197658
Bajo: 197127


##### Hallazgos de consistencia nivel de satisfacción

Se identificó una alta inconsistencia en la estandarización de los datos de la columna `nivel_satisfaccion`. La variable presenta un total de **15 variantes de formato distintas** para representar únicamente 3 categorías reales (*Alto*, *Medio*, *Bajo*), además de un porcentaje de registros nulos.

Se identificaron las siguientes inconsistencias:

1. **Espacios en blanco adicionales (*Trailing/Leading spaces*):** Presencia de espacios al inicio (ej. `' Alto'`) o al final (ej. `'Medio '`), provocando que Pandas los interprete como etiquetas independientes.
2. **Uso inconsistente de mayúsculas y minúsculas:**
   * Totalmente en minúsculas: `alto`, `medio`, `bajo`
   * Sostenidas en mayúsculas: `ALTO`, `MEDIO`, `BAJO`
   * Formato título: `Alto`, `Medio`, `Bajo`
3. **Valores nulos (`NaN`):** Se registraron **46,469** datos faltantes, lo cual equivale al **7.26%** del total de la muestra (640,000 registros).

Los datos agrupados quedaron de esta manera:

| Categoría Unificada | Frecuencia Absoluta | Proporción (%) |
| :--- | :---: | :---: |
| **Alto** | 198,746 | 31.05% |
| **Medio** | 197,658 | 30.88% |
| **Bajo** | 197,127 | 30.80% |
| **Nulos (`NaN`)** | 46,469 | 7.26% |
| **Total** | **640,000** | **100.00%** |

### **2.4. Validez**

In [3]:
df[~df['correo_cliente'].str.contains('@', na = False)]

,id_pedido,id_cliente,fecha_pedido,canal_compra,metodo_pago,ciudad_tienda,codigo_postal,categoria_producto,producto,precio_unitario,unidades_vendidas,monto_compra,edad_cliente,correo_cliente,nivel_satisfaccion,nivel_lealtad,comentario_cliente,peso_pedido_kg,fecha_actualizacion_stock,fuga_cliente
12,P500774,C159046,1/1/2026,App,efectivo,Bucaramanga,6800101,Moda,Camisa de lino,344500.0,1,344500.0,53,NaN,Bajo,Oro,"Excelente calidad, volvería a comprar",13.4,2026-07-05,1
13,P566871,C103541,04/01/2026,Marketplace,PayPal,Cúcuta,5400109,Electrónica,Parlante bluetooth,154600.0,3,463800.0,57,NaN,Bajo,Bronce,"Excelente calidad, volvería a comprar",10.9,2026-07-16,0
18,P693809,C155675,19/2/2026,App,Efectivo,Villavicencio,5000103,Juguetería,Muñeca articulada,316200.0,1,316200.0,56,cliente593809.hotmail.com,bajo,Oro,"Excelente calidad, volvería a comprar",11.6,2023-10-03,0
25,P683717,C114516,06/10/2025,web,tarjeta,Barranquilla,800101,Electrónica,Teclado mecánico,305800.0,1,305800.0,65,cliente583717.outlook.com,Bajo,Platino,"Todo perfecto, llegó a tiempo",10.6,2026-02-26,0
33,P635537,C130110,19/12/2025,App,Tarjeta,Manizales,1700103,DEPORTES,Pesas ajustables,137200.0,2,274400.0,52,NaN,Alto,Oro,Me costó encontrar el producto en la app,8.3,2026-07-27,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
639982,P231923,C27475,20/01/2025,Marketplace,Tarjeta,ibague,7300101,Belleza,Secador de pelo,213800.0,5,1069000.0,200,NaN,Bajo,Bronce,Llegó incompleto el pedido,12.3,2027-08-11,1
639986,P307845,C136455,2026-05-14,Tienda,PayPal,CTG,1300105,Juguetería,Carro a control remoto,303700.0,1,303700.0,45,NaN,Alto,Plata,Llegó incompleto el pedido,14.4,2026-04-03,0
639988,P238956,C74804,12/04/2025,WEB,PayPal,VILLAVICENCIO,5000103,Moda,Camisa de lino,470800.0,2,941600.0,20,NaN,Medio,Oro,Pésima experiencia con el empaque,10.2,2026-06-29,0
639993,P543968,C141410,2026-02-18,App,Transferencia,Villavicencio,5000101,electronica,Monitor 24 pulgadas,111400.0,1,111400.0,56,NaN,Alto,Bronce,Rápido y sin problemas,12.0,2026-06-18,0


In [5]:
fechas_convertidas = pd.to_datetime(df['fecha_pedido'], errors='coerce')
fechas_invalidas = df[fechas_convertidas.isna() | (fechas_convertidas > pd.Timestamp.today())]
fechas_invalidas

C:\Users\juane\AppData\Local\Temp\ipykernel_12968\3961334209.py:1: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  fechas_convertidas = pd.to_datetime(df['fecha_pedido'], errors='coerce')


,id_pedido,id_cliente,fecha_pedido,canal_compra,metodo_pago,ciudad_tienda,codigo_postal,categoria_producto,producto,precio_unitario,unidades_vendidas,monto_compra,edad_cliente,correo_cliente,nivel_satisfaccion,nivel_lealtad,comentario_cliente,peso_pedido_kg,fecha_actualizacion_stock,fuga_cliente
1,P677836,C84260,2026-03-05,Web,Efectivo,MZL,1700101,Electrónica,Parlante bluetooth,195700.0,2,391400.0,73,cliente577836@yahoo.es,Medio,Bronce,El producto llegó dañado,12.9,2026-06-10,0
2,P387869,C164166,2025-03-13,Tienda,PayPal,Cúcuta,5400101,Juguetería,Rompecabezas 1000 piezas,597600.0,-3,-1792800.0,32,cliente287869@novamarket.com.co,Bajo,Oro,El domiciliario fue muy amable,14.3,2026-02-26,0
5,P465514,C102032,2025-08-29,Tienda,Efectivo,MDE,500102,Deportes,Balón de fútbol,626800.0,2,1253600.0,33,cliente365514@yahoo.es,Alto,Platino,Llegó incompleto el pedido,12.7,2027-07-21,0
7,P583062,C94873,19/08/2027,TIENDA,PayPal,Medellín,500103,Deportes,Bicicleta estática,190900.0,-1,-190900.0,26,cliente483062@outlook.com,Alto,Platino,NaN,8.4,2024-03-14,0
9,P622879,C155042,2025-02-23,Marketplace,efectivo,CALI,7600101,Electrónica,Audífonos inalámbricos,-124300.0,3,-372900.0,67,cliente522879@novamarket.com.co,Alto,Oro,No era lo que esperaba,15.0,2026-05-08,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
639989,P683071,C28408,2025-10-12,marketplace,Transferencia,Bucaramanga,6800112,Deportes,Colchoneta de yoga,NaN,4,NaN,29,cliente583071@hotmail.com,Alto,Platino,"Todo perfecto, llegó a tiempo",1237.6,2026-07-28,0
639990,P257363,C23142,2025-07-30,Web,tarjeta,Medellín,500101,Electrónica,Monitor 24 pulgadas,416400.0,4,1665600.0,68,cliente157363@gmail.com,Alto,Plata,NaN,12.5,2026-05-26,0
639991,P711828,C133738,2025-11-29,app,PayPal,CALI,7600129,JUGUETERÍA,Carro a control remoto,220200.0,3,660600.0,27,cliente611828@hotmail.com,Bajo,Bronce,Cumplió con lo prometido,9.5,2026-07-15,1
639993,P543968,C141410,2026-02-18,App,Transferencia,Villavicencio,5000101,electronica,Monitor 24 pulgadas,111400.0,1,111400.0,56,NaN,Alto,Bronce,Rápido y sin problemas,12.0,2026-06-18,0


In [8]:
cant_correos = len(df[~df['correo_cliente'].str.contains('@', na=False)])
cant_fechas = len(df[fechas_convertidas.isna() | (fechas_convertidas > pd.Timestamp.today())])

print(f"Correos inválidos: {cant_correos}")
print(f"Fechas inválidas/futuras: {cant_fechas}")

Correos inválidos: 156844
Fechas inválidas/futuras: 239538


##### Hallazgos de validez para correos y fechas

En cuanto a la validez de los datos, encontramos registros con formatos incorrectos que impiden procesar la información. Lo primero es que en la columna fecha_pedido, hay 239538 fechas con un formato inválido que no se pueden leer correctamente para saber cuándo se realizo dicha venta o tienen una fecha que está en el futuro, algo completamente ilogico en cualquier tipo de registro. Además, encontramos 156844 correos electrónicos que están mal escritos porque les falta el arroba (@) o el punto (.).

### **2.5. Unicidad**

La unicidad evalúa si cada registro que debería representar una entidad diferente aparece una sola vez en el dataset.

Para esta auditoría se utiliza `id_pedido` como identificador del pedido, ya que cada pedido debería tener un identificador único en toda la base de datos.

In [2]:
# Duplicados por id_pedido
duplicados_id = df.duplicated(subset=['id_pedido'], keep='first')
cantidad_duplicados = duplicados_id.sum()
porcentaje_duplicados = cantidad_duplicados / len(df) * 100
ids_unicos = df['id_pedido'].nunique()

# Duplicados de fila completa
duplicados_completos = df.duplicated(keep='first').sum()

# Registros con id_pedido repetido (para inspección)
registros_ids_repetidos = df[
    df.duplicated(subset=['id_pedido'], keep=False)
].sort_values('id_pedido')

print("Registros totales:", len(df))
print("ID de pedidos únicos:", ids_unicos)
print("Registros duplicados por id_pedido:", cantidad_duplicados)
print("Filas completamente duplicadas:", duplicados_completos)
print("Porcentaje de duplicados:", round(porcentaje_duplicados, 2), "%")

print("\nEjemplos de registros con id_pedido repetido:")
display(registros_ids_repetidos[['id_pedido', 'id_cliente', 'fecha_pedido']].head(10))

Registros totales: 640000
ID de pedidos únicos: 620000
Registros duplicados por id_pedido: 20000
Filas completamente duplicadas: 20000
Porcentaje de duplicados: 3.12 %

Ejemplos de registros con id_pedido repetido:


,id_pedido,id_cliente,fecha_pedido
172693,P100005,C164871,2025-01-28
174646,P100005,C164871,2025-01-28
418817,P100010,C76626,17/8/2025
368632,P100010,C76626,17/8/2025
168620,P100023,C46545,00/00/0000
103928,P100023,C46545,00/00/0000
205691,P100158,C14684,2025-10-12
46328,P100158,C14684,2025-10-12
57401,P100163,C88511,5/8/2025
146,P100163,C88511,5/8/2025


##### Hallazgos de unicidad

Al revisar la unicidad usando `id_pedido` como identificador de cada pedido, encontramos que de los **640.000 registros** solo hay **620.000 IDs únicos**, lo que significa que hay **20.000 registros duplicados (3,12 % del dataset)**.

Para verificar si se trataba solo de una reutilización del identificador o de filas totalmente iguales, revisamos también los duplicados de fila completa y encontramos que **también son 20.000 filas exactamente iguales**. Es decir, esos pedidos están registrados dos veces con toda su información idéntica.

Este tipo de duplicados puede inflar cualquier análisis posterior: se contarían pedidos, unidades vendidas e ingresos por partida doble. Lo recomendable es conservar una sola ocurrencia de cada registro y establecer un control sobre `id_pedido` para que no se pueda registrar dos veces el mismo pedido.

### **2.6. Oportunidad**

La oportunidad evalúa si la información se encuentra disponible y actualizada en el momento adecuado para su uso.

En esta sección se revisan dos aspectos:

- `fecha_pedido`, para identificar registros con fechas futuras que no deberían existir en un dataset operacional.
- `fecha_actualizacion_stock`, para determinar qué tan actualizada está la información de inventario.

Para el stock se establece un umbral de **90 días**: un registro con más de tres meses sin actualización deja de representar bien la disponibilidad actual de un producto. La fecha de corte de la auditoría es el **11 de agosto de 2026**.

In [3]:
# Conversión robusta de fecha_pedido (maneja formatos mezclados)
fecha_txt = df['fecha_pedido'].astype('string').str.strip()

formato_iso = fecha_txt.str.match(r'^\d{4}-\d{1,2}-\d{1,2}$', na=False)
formato_dmy = fecha_txt.str.match(r'^\d{1,2}/\d{1,2}/\d{4}$', na=False)

fecha_convertida = pd.Series(pd.NaT, index=df.index, dtype='datetime64[ns]')
fecha_convertida.loc[formato_iso] = pd.to_datetime(fecha_txt.loc[formato_iso], format='%Y-%m-%d', errors='coerce')
fecha_convertida.loc[formato_dmy] = pd.to_datetime(fecha_txt.loc[formato_dmy], format='%d/%m/%Y', errors='coerce')

# Fecha de referencia de la auditoría
fecha_corte = pd.Timestamp('2026-08-11')

# 1. Pedidos con fecha futura
pedidos_futuros = fecha_convertida.notna() & (fecha_convertida > fecha_corte)
cantidad_pedidos_futuros = pedidos_futuros.sum()
porcentaje_pedidos_futuros = cantidad_pedidos_futuros / len(df) * 100

# 2. Stock desactualizado (> 90 días)
fecha_stock = pd.to_datetime(df['fecha_actualizacion_stock'], errors='coerce')
umbral_dias = 90
fecha_limite = fecha_corte - pd.Timedelta(days=umbral_dias)

stock_desactualizado = fecha_stock.notna() & (fecha_stock < fecha_limite)
cantidad_stock_desactualizado = stock_desactualizado.sum()
porcentaje_stock_desactualizado = cantidad_stock_desactualizado / len(df) * 100

# 3. Stock con fecha futura
stock_futuro = fecha_stock.notna() & (fecha_stock > fecha_corte)
cantidad_stock_futuro = stock_futuro.sum()
porcentaje_stock_futuro = cantidad_stock_futuro / len(df) * 100

resultados_oportunidad = pd.DataFrame({
    'Problema': [
        'Pedidos con fecha futura',
        'Stock con más de 90 días sin actualizar',
        'Actualización de stock futura'
    ],
    'Registros afectados': [
        cantidad_pedidos_futuros,
        cantidad_stock_desactualizado,
        cantidad_stock_futuro
    ],
    'Porcentaje': [
        round(porcentaje_pedidos_futuros, 2),
        round(porcentaje_stock_desactualizado, 2),
        round(porcentaje_stock_futuro, 2)
    ]
})

print("Fecha de corte:", fecha_corte.date())
print("Umbral de oportunidad:", umbral_dias, "días")
print("Fecha límite para stock actualizado:", fecha_limite.date())
print()
print(resultados_oportunidad.to_string(index=False))

print("\nRango de fecha_pedido válida:")
print("Mínima:", fecha_convertida.min())
print("Máxima:", fecha_convertida.max())

print("\nRango de fecha_actualizacion_stock:")
print("Mínima:", fecha_stock.min())
print("Máxima:", fecha_stock.max())

Fecha de corte: 2026-08-11
Umbral de oportunidad: 90 días
Fecha límite para stock actualizado: 2026-05-13

                               Problema  Registros afectados  Porcentaje
               Pedidos con fecha futura                16485        2.58
Stock con más de 90 días sin actualizar               350611       54.78
          Actualización de stock futura                21722        3.39

Rango de fecha_pedido válida:
Mínima: 2025-01-01 00:00:00
Máxima: 2027-12-28 00:00:00

Rango de fecha_actualizacion_stock:
Mínima: 2023-01-01 00:00:00
Máxima: 2027-12-28 00:00:00


##### Hallazgos de oportunidad

Tomando como fecha de corte el **11 de agosto de 2026**, encontramos que en `fecha_pedido` hay **16.485 registros con fechas futuras (2,58 % del dataset)**. Son pedidos que, según la fecha de la auditoría, ni siquiera deberían haber ocurrido — algunos llegan hasta diciembre de 2027.

Para `fecha_actualizacion_stock` fijamos un umbral de 90 días, porque NovaMarket usa la información de inventario para decisiones operativas y un registro con más de tres meses sin actualización deja de reflejar bien la disponibilidad real del producto. Bajo ese criterio, encontramos que **350.611 registros (54,78 %) tienen el stock desactualizado**, lo cual es más de la mitad del dataset y es de lejos el problema más grave de esta dimensión.

Además, aparecen **21.722 registros (3,39 %) con fecha de actualización de stock en el futuro**, lo cual no debería ser posible.

El problema principal está claramente en la actualización del inventario: cualquier decisión de reposición o reporte de disponibilidad construida sobre estos datos va a estar sesgada. Se recomienda establecer una frecuencia máxima de actualización del inventario y controles que impidan registrar fechas posteriores a la fecha real de la operación.

## **3. Catálogo de Problemas**

A continuación se consolidan todos los hallazgos de las seis dimensiones auditadas sobre el dataset de NovaMarket (640.000 registros × 20 columnas).

| Problema | Dimensión | Columnas afectadas | Registros afectados | Impacto |
| :--- | :--- | :--- | :--- | :--- |
| **Datos faltantes / nulos** | Completitud | correo_cliente, comentario_cliente, nivel_satisfaccion, ciudad_tienda, precio_unitario, monto_compra | correo_cliente: 105.312 (16,46 %)<br>comentario_cliente: 90.806 (14,19 %)<br>nivel_satisfaccion: 46.469 (7,26 %)<br>ciudad_tienda: 29.928 (4,68 %)<br>precio_unitario: 21.640 (3,38 %)<br>monto_compra: 21.640 (3,38 %) | **Alto** |
| **Valores numéricos imposibles** | Exactitud | precio_unitario, unidades_vendidas, edad_cliente | precio_unitario < 0: 19.629 (3,07 %)<br>precio_unitario > 10 M: 6.164 (0,96 %)<br>unidades_vendidas < 0: 18.550 (2,90 %)<br>unidades_vendidas > 100: 18.526 (2,89 %)<br>edad_cliente fuera de [0,120]: 27.829 (4,35 %) | **Alto** |
| **Inconsistencia por formato** | Consistencia | ciudad_tienda, categoria_producto, canal_compra, metodo_pago, nivel_satisfaccion | ciudad_tienda: 314.864 (49,20 %)<br>categoria_producto: 304.697 (47,61 %)<br>canal_compra: 154.781 (24,18 %)<br>metodo_pago y nivel_satisfaccion con patrones similares | **Alto** |
| **Formato de correo y fecha inválidos** | Validez | correo_cliente, fecha_pedido | correo_cliente sin `@` o mal formado: 51.532 (8,05 %)<br>fecha_pedido inválidas o no parseables: 239.538 (37,43 %) | **Alto** |
| **Filas o claves duplicadas** | Unicidad | id_pedido, filas completas | id_pedido duplicados: 20.000 (3,12 %)<br>Filas completamente duplicadas: 20.000 (3,12 %) | **Medio** |
| **Fechas futuras respecto al análisis** | Oportunidad | fecha_pedido, fecha_actualizacion_stock | fecha_pedido futura: 16.485 (2,58 %)<br>Stock con fecha futura: 21.722 (3,39 %) | **Medio** |
| **Inventario desactualizado (> 90 días)** | Oportunidad | fecha_actualizacion_stock | 350.611 registros (54,78 %) | **Alto** |

## **4. Priorización**

**1. Inventario desactualizado (Oportunidad).** Más de la mitad del dataset (54,78 %) tiene el stock sin actualizar hace más de 90 días. Cualquier reporte de disponibilidad, decisión de reposición o análisis de rotación construido sobre estos datos va a estar sesgado. Es el problema más masivo del dataset.

**2. Inconsistencias de formato en categóricas (Consistencia).** Casi la mitad de los registros de `ciudad_tienda` (49,20 %) y `categoria_producto` (47,61 %) tienen variantes de escritura de la misma categoría (mayúsculas, tildes, espacios, abreviaturas). Sin homologar, cualquier análisis de ventas por ciudad o categoría reporta dobles: por ejemplo, "Bogotá" y "bogota" se cuentan como dos ciudades diferentes.

**3. Fechas y correos inválidos (Validez).** El 37,43 % de `fecha_pedido` no se puede parsear correctamente y el 8,05 % de los correos no tienen formato válido. Sin fechas confiables no se puede saber cuándo se hizo cada venta; sin correos válidos no llegan facturaciones ni notificaciones al cliente.

**4. Datos faltantes (Completitud).** Faltan el 16,46 % de correos y el 3,38 % de precios/montos. Los precios faltantes rompen los cálculos de ingresos totales; los correos faltantes limitan la comunicación con el cliente.

**5. Valores numéricos imposibles (Exactitud).** Precios negativos, unidades negativas, edades fuera del rango humano y precios de más de 10 millones. Cada uno afecta menos del 5 % individualmente, pero en conjunto distorsionan las agregaciones (medias, sumas, ingresos totales).

**6. Duplicados de pedido (Unicidad).** 20.000 pedidos aparecen dos veces con toda la información idéntica (3,12 %). Genera doble conteo de ventas, unidades e ingresos.

**7. Pedidos con fecha futura (Oportunidad).** 16.485 registros (2,58 %) tienen fechas posteriores a la fecha de corte, algunas hasta 2027. Toca decidir si se corrigen o se eliminan.

## **5. Recomendaciones**

**Fechas.** Convertir `fecha_pedido` a formato estándar (`YYYY-MM-DD`) haciendo un parseo tolerante a los formatos mezclados (`DD/MM/YYYY`, `D/M/YYYY`, `YYYY-MM-DD`). Marcar como nulas las fechas que no se puedan interpretar y revisar caso por caso las que quedan en el futuro. Para `fecha_actualizacion_stock`, establecer un proceso periódico de actualización y un control que impida registrar fechas posteriores a la fecha de operación.

**Correos.** Para los faltantes, asignar un placeholder como `'sin_correo'` o dejarlos nulos según la política del negocio. Para los inválidos (sin `@` o mal formados), enviarlos a un flujo de revisión manual antes de usarlos en campañas de comunicación.

**Numéricos.** Pasar a positivo los valores negativos de `precio_unitario`, `unidades_vendidas` y `edad_cliente` cuando el signo negativo sea claramente un error de captura. Para `precio_unitario` mayor a 10 millones y `unidades_vendidas` mayor a 100, revisar caso por caso — pueden ser errores de digitación (ceros de más) o casos legítimos. Los precios faltantes se pueden imputar por la mediana del producto o categoría.

**Categóricas.** Homologar `ciudad_tienda`, `categoria_producto`, `canal_compra`, `metodo_pago` y `nivel_satisfaccion` quitando espacios al inicio y final, normalizando a minúsculas y luego mapeando cada variante a su categoría estándar. También ajustar el `codigo_postal` según la ciudad correspondiente después de la homologación.

**Unicidad.** Eliminar los 20.000 pedidos duplicados conservando una sola ocurrencia por `id_pedido` y establecer una restricción única en la fuente para que no se puedan volver a registrar.